### PR TITLE
Added snowball

### PR DESCRIPTION
**Pull request by mauro-balades**

### CODE REVIEW

**Comment by mauro-balades** [comment]
how can I transfere snowball/objects/* to /usr/local/lib/*

**Comment by jeremy-rifkin** [comment]
Everything needs to live in /opt/compiler-explorer/snowball{version} (or whatnot). We have hundreds of compilers and versions, if everything tried to put stuff in /usr we’d surely have issues :)

if you test locally, does snowball install properly into /opt?

**Comment by mauro-balades** [comment]
It did install but it requires libSnowball.so that's inside the "objects" folder inside the tar. Is there a way I can manage to fix the LD_PATH or smth?

**Comment by mattgodbolt** [comment]
> It did install but it requires libSnowball.so that's inside the "objects" folder inside the tar. Is there a way I can manage to fix the LD_PATH or smth?

The usual way is to compile with an `rpath` that looks at an `$ORIGIN`-relative directory for dependent libraries.

**Comment by mattgodbolt** [comment]
How is `snowball`-built? If it's using CMake then there's probably an easy way to do this.

**Comment by mattgodbolt** [comment]
I found it: in CMake if the installation directory for your libs is `../lib` compared to the install then:
```
  set(CMAKE_INSTALL_RPATH $ORIGIN/../lib)
```
should do the trick.

**Comment by mattgodbolt** [comment]
I also took a quick look at the archive you built:

```
$ curl -sL https://github.com/snowball-lang/snowball/releases/download/v0.0.7/snowball-linux-x86_64.tar.gz | tar ztv
drwxr-xr-x build/build       0 2023-07-09 04:02 release/
drwxr-xr-x build/build       0 2023-07-09 04:02 release/objects/
-rwxr-xr-x build/build   28832 2023-07-09 03:59 release/objects/libSnowballRuntime.so
...
```

It looks like `objects` is the place you install to, which is unusual? Do you actually `make install` before archiving?

I'd suggest you don't archive up the `release` directory but instead either drop the `release` prefix or better still make it `snowball-0.7` or something.

I suspect though you're just archiving the built directory and not the installed directory, which is going to break my earlier suggestion to set `CMAKE_INSTALL_RPATH` (which requires that you `make install`)

**Comment by mauro-balades** [comment]
What's up, sorry for not answering...

If I'm understanding correctly, you are suggesting I do a separate release for compiler explorer that instead of searching for builtin modules in ~/.snowball and libs in /usr/lib etc, change it so that it self-encapsulates everything into a single folder?

**Comment by mauro-balades** [comment]
> I found it: in CMake if the installation directory for your libs is `../lib` compared to the install then:
> 
> ```
>   set(CMAKE_INSTALL_RPATH $ORIGIN/../lib)
> ```
> 
> should do the trick.

is $ORIGIN built in?

**Comment by Endilll** [comment]
Yes, it's something dynamic linker (ld.so) understands.
You can read more [here](https://man7.org/linux/man-pages/man8/ld.so.8.html).

**Comment by mauro-balades** [comment]
thanks!

**Comment by mauro-balades** [comment]
The output for `ce_install` was:

```
2023-07-09 19:17:23,045 lib.installation_context INFO     Making uncached requests
2023-07-09 19:17:23,045 lib.ce_install  INFO     Creating multiprocessing pool with 8 workers
Installing compilers/snowball 0.0.7
2023-07-09 19:17:27,566 lib.installation_context INFO     Fetching https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz (31754290 bytes)
2023-07-09 19:17:32,870 lib.installation_context INFO     66.0% of https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz...
2023-07-09 19:17:35,520 lib.installation_context INFO     100% of https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz
2023-07-09 19:17:35,521 lib.installation_context INFO     Piping to tar zxf - --strip-components 1
2023-07-09 19:17:36,331 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/b2fcc154-f2e7-44fd-94ec-ec43976116c2/snowball-0.0.7) to final destination (/opt/compiler-explorer/snowball-0.0.7)
2023-07-09 19:17:36,368 lib.ce_install  INFO     compilers/snowball 0.0.7 installed OK
1 packages installed OK, 0 skipped, and 0 failed installation
```

I believe that is a... successful installation 😎

**Comment by mattgodbolt** [comment]
Amazing! Will merge and try on our admin node

File: bin/yaml/snowball.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    compression: gz
+    dir: snowball-{name}
+    create_untar_dir: true
+    strip_components: 1
+    build: snowball-{name}-release
+    toolchain: cmake
+    check_exe: bin/snowball --version
+    targets:
+      - name: '0.0.7'
+        url: https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz
```
Comment (mattgodbolt) [inline bin/yaml/snowball.yaml:13]: Ideally the URL should be related to the `{name}` but as this has `hotfix` in the middle I guess that's just a special case for this release?

**Review by mattgodbolt — APPROVED** [review]

**Comment by mattgodbolt** [comment]
No luck, sadly:
```bash
Jul/09 19:56 admin-node~/infra (main|✔) $ sudo ce_install install snowball
2023-07-09 19:56:59,134 lib.installation_context INFO     Making uncached requests
2023-07-09 19:56:59,135 lib.ce_install  INFO     Creating multiprocessing pool with 2 workers
Installing compilers/snowball 0.0.7
2023-07-09 19:57:04,380 lib.installation_context INFO     Fetching https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz (31760671 bytes)
2023-07-09 19:57:04,562 lib.installation_context INFO     100% of https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz
2023-07-09 19:57:04,563 lib.installation_context INFO     Piping to tar zxf - --strip-components 1
2023-07-09 19:57:06,230 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/981ba9b7-dabc-470c-bc9d-806cf2fd1765/snowball-0.0.7) to final destination (/opt/compiler-explorer/snowball-0.0.7)
/opt/compiler-explorer/snowball-0.0.7/bin/snowball: error while loading shared libraries: libSnowball.so: cannot open shared object file: No such file or directory
2023-07-09 19:57:06,330 lib.ce_install  ERROR    compilers/snowball 0.0.7 installed OK, but doesn't appear as installed after
0 packages installed OK, 0 skipped, and 1 failed installation
Failed:
  compilers/snowball 0.0.7
```

```
Jul/09 19:57 admin-node~/infra (main|✔) 8.7s $ ldd /opt/compiler-explorer/snowball-0.0.7/bin/snowball
        linux-vdso.so.1 (0x00007ffe639f4000)
        libSnowball.so => not found
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc24a600000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc24a918000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc24a200000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc24a831000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc24a962000)
Jul/09 19:57 admin-node~/infra (main|✔) $ readelf -d /opt/compiler-explorer/snowball-0.0.7/bin/snowball

Dynamic section at offset 0x1ab20 contains 31 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libSnowball.so]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000001d (RUNPATH)            Library runpath: [/ws:::::::::::]
 0x000000000000000c (INIT)               0x6000
```

the "Library runpath" is not set to `$ORIGIN/..lib`, it's set to `/ws:::::::::`

**Comment by mattgodbolt** [comment]
For comparison this is what a clang tool looks like:
```
Jul/09 19:58 admin-node~/infra (main|✔) $ readelf -d /opt/compiler-explorer/clang-16.0.0/bin/llvm-objdump

Dynamic section at offset 0x21f9318 contains 35 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN/../lib]
```

File: bin/yaml/snowball.yaml (hunk 1 above)
Comment (mauro-balades) [reply bin/yaml/snowball.yaml:13]: Yeah, this is a one time case because version 0.0.7 did not include Ce-specific bi aries.

**Comment by mauro-balades** [comment]
I don't get it, this is the code:

```cmake
if (DEFINED ENV{BUILD_FOR_CE})
  message(STATUS "Building for CE")
  set(CMAKE_INSTALL_RPATH $ORIGIN/../lib)
endif()
```

**Comment by mattgodbolt** [comment]
And do you actually install this for archiving?

**Comment by mauro-balades** [comment]
does it need to be before or after the `project()` directive?

**Comment by mattgodbolt** [comment]
It doesn't matter where it's defined.

But your build scripts dont' actually install anything I think: Looking at: https://github.com/snowball-lang/snowball/blob/dev/build_scripts/build-snowball.sh you don't actually use cmake's installation system, you are manually copying things around. That means the rpath stuff won't be applied.

**Comment by mauro-balades** [comment]
ohh... how could it be installed into the release directory then?

**Comment by mattgodbolt** [comment]
The "cmake way" is to add a bunch of `install()` definitions that explain how to install your project. This will copy all the libraries you built into relevant places, and then the "norma" thing for us is to "install" to a temp directory and then tarball that up.

The cmake installation process understands rpaths and knows how to patch the executables and libraries to use them appropriatley.

It's normal to leave the CMAKE_INSTALL_RPATH set, it's not something that necessarily needs to be guarded behind a IF in the cmake. But it's platform-specific for linux, so you might want to do it anyway.

A workaround would be to use a special "build for CE" that manually rewrites the rpath with `elfedit` or similar.

**Comment by mauro-balades** [comment]
I think I also need to add:

```cmake
set( CMAKE_BUILD_WITH_INSTALL_RPATH FALSE ) 
```

**Comment by mattgodbolt** [comment]
> ohh... how could it be installed into the release directory then?

with something like:

```
$ make DESTDIR=/wherever/you/want/the/release/to/go install
```

lots more info at 
https://cmake.org/cmake/help/latest/command/install.html and https://cmake.org/cmake/help/v3.0/variable/CMAKE_INSTALL_PREFIX.html

it's a huge topic and I can't pretend to be an expert, especially with cross-platform. But manually copying things around in a shell script seems like the worst of all worlds if you're using CMake anyway :)

**Comment by mauro-balades** [comment]
I think I might do `patchelf --set-rpath '$ORIGIN/../lib' bin/snowball` for ce-builds

**Comment by mattgodbolt** [comment]
you do whatever works for you

**Comment by mauro-balades** [comment]
The binaries will be automatically updated when this finishes 👍🏽 

https://github.com/snowball-lang/snowball/actions/runs/5501802097/jobs/10025641219

**Comment by mattgodbolt** [comment]
Got it - will watch :)

**Comment by mattgodbolt** [comment]
```
  File "/home/ubuntu/infra/bin/lib/installation_context.py", line 118, in fetch_to
    raise FetchFailure(f"Fetch failure for {url}: {request}")
lib.installation_context.FetchFailure: Fetch failure for https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz: <Response [404]>

0 packages installed OK, 0 skipped, and 1 failed installation
Failed:
  compilers/snowball 0.0.7

```

**Comment by mauro-balades** [comment]
yes, I deleted the hotfix release and created a new one. It will work once it finishes compiling (but ofc, it's a c++ project so it might take 2.6 years)

**Comment by mattgodbolt** [comment]
But it had finished? The link you sent had completed:
[Build ce-specific (x86_64) [ce-specific]](https://github.com/snowball-lang/snowball/actions/runs/5501802097/jobs/10025641219#logs)
cancelled 1 minute ago in 1m 29s

**Comment by mattgodbolt** [comment]
oh...you canceled it?

**Comment by mauro-balades** [comment]
https://github.com/snowball-lang/snowball/actions/runs/5501811652/jobs/10025656229

**Comment by mauro-balades** [comment]
it will take max 2 minutes left 👍🏽

**Comment by mauro-balades** [comment]
@mattgodbolt you can try now... please god make it work 🙏

**Comment by mattgodbolt** [comment]
```
$ sudo ce_install install snowball
2023-07-09 20:41:16,170 lib.installation_context INFO     Making uncached requests
2023-07-09 20:41:16,171 lib.ce_install  INFO     Creating multiprocessing pool with 2 workers
Installing compilers/snowball 0.0.7
2023-07-09 20:41:21,360 lib.installation_context INFO     Fetching https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz (31760960 bytes)
2023-07-09 20:41:21,558 lib.installation_context INFO     100% of https://github.com/snowball-lang/snowball/releases/download/v0.0.7-hotfix/snowball-ce-specific-x86_64.tar.gz
2023-07-09 20:41:21,558 lib.installation_context INFO     Piping to tar zxf - --strip-components 1
2023-07-09 20:41:23,176 lib.installation_context INFO     Moving from staging (/opt/compiler-explorer/staging/a7a22816-68b8-455d-950e-225380b547a4/snowball-0.0.7) to final destination (/opt/compiler-explorer/snowball-0.0.7)
2023-07-09 20:41:24,922 lib.ce_install  INFO     compilers/snowball 0.0.7 installed OK
1 packages installed OK, 0 skipped, and 0 failed installation
```
hooray

**Comment by mauro-balades** [comment]
yay!!!

**Comment by mattgodbolt** [comment]
Just as a final check I will run a test on an actual CE instance (which is more cut down than the admin node where I installed)
and...
YAY
```
ubuntu@ip-172-30-5-8:~$ /opt/compiler-explorer/snowball-0.0.7/bin/snowball --help
Snowball (C) MIT
Usage: snowball [command] [options]

Snowball Commands:
  --version [-v]   - builds the current snowball project
  build            - builds the current snowball project
  run              - build and execute the current snowball project
  test             - execute tests
  init             - create a new snowball project (in current directory)
  help             - prints out help

Try `snowball help [command]` for more information.
```

**Comment by mauro-balades** [comment]
im so happy rn, thanks a lot matt

**Comment by mauro-balades** [comment]
one question, when will it be updated on the website?

**Comment by mattgodbolt** [comment]
Hopefully later today or tomorrow: I'll have to look at the compiler-explorer change too and merge that and then do a site update.

**Pull request merged**